src/Components/Server/test/Circuits/RevalidatingServerAuthenticationStateProvider.cs

```diff
@@ -0,0 +1,194 @@
+        }
+
+        [Fact]
+        public async Task IfValidateAuthenticationStateAsyncThrows_ForcesSignOut()
+        {
+            // Arrange
+            using var provider = new TestRevalidatingServerAuthenticationStateProvider(
+                TimeSpan.FromMilliseconds(50));
+            provider.SetAuthenticationState(CreateAuthenticationStateTask("test user"));
+            provider.NextValidationResult = Task.FromException<bool>(new InvalidTimeZoneException());
+
+            var newAuthStateNotificationTcs = new TaskCompletionSource<Task<AuthenticationState>>();
+            provider.AuthenticationStateChanged += newStateTask => newAuthStateNotificationTcs.SetResult(newStateTask);
+
+            // Act
+            var newAuthStateTask = await newAuthStateNotificationTcs.Task;
+            var newAuthState = await newAuthStateTask;
+
+            // Assert
+            Assert.False(newAuthState.User.Identity.IsAuthenticated);
+
+            // Assert: no longer revalidates
+            await Task.Delay(200);
+            Assert.Single(provider.RevalidationCallLog);
+        }
+
+        [Fact]
+        public async Task IfHostSuppliesNewAuthenticationState_RestartsRevalidationLoop()
+        {
+            // Arrange
+            using var provider = new TestRevalidatingServerAuthenticationStateProvider(
+                TimeSpan.FromMilliseconds(50));
+            provider.SetAuthenticationState(CreateAuthenticationStateTask("test user"));
+            provider.NextValidationResult = Task.FromResult(true);
+            await provider.NextValidateAuthenticationStateAsyncCall;
+            Assert.Collection(provider.RevalidationCallLog,
+                authState => Assert.Equal("test user", authState.User.Identity.Name));
+
+            // Act/Assert 1: Can become signed out
+            // Doesn't revalidate unauthenticated states
+            provider.SetAuthenticationState(CreateAuthenticationStateTask(null));
+            await Task.Delay(200);
+            Assert.Empty(provider.RevalidationCallLog.Skip(1));
+
+            // Act/Assert 2: Can become a different user; resumes revalidation
+            provider.SetAuthenticationState(CreateAuthenticationStateTask("different user"));
+            await provider.NextValidateAuthenticationStateAsyncCall;
+            Assert.Collection(provider.RevalidationCallLog.Skip(1),
+                authState => Assert.Equal("different user", authState.User.Identity.Name));
+        }
+
+        [Fact]
+        public async Task StopsRevalidatingAfterDisposal()
+        {
+            // Arrange
+            using var provider = new TestRevalidatingServerAuthenticationStateProvider(
+                TimeSpan.FromMilliseconds(50));
+            provider.SetAuthenticationState(CreateAuthenticationStateTask("test user"));
+            provider.NextValidationResult = Task.FromResult(true);
+
+            // Act
+            ((IDisposable)provider).Dispose();
+            await Task.Delay(200);
```

Comment:
Think about providing access to the actual validation task somehow so we can await it. This kind of thing ends up being flake eventually usually. Yeah I know that 200 > 50

Reply:
That's what I have done in the cases where there is a validation task that can be awaited (it's `NextValidateAuthenticationStateAsyncCall`).

However for the tests where we're asserting that no further calls get made, such as this one, there's no validation task to await.

Reply:
Without changing the public API in an unnatural test-centric manner (e.g., passing in some `IDelayTaskProvider`) I think this is the best we can do, so I propose leaving it. I haven't observed any hints of flakiness when running it locally so far.

If this does become a problem in the future, we do have the option to refactor without breaking any compatibility, e.g.:

 * Add a new `internal` property of type `IDelayTaskProvider`
 * If that is assigned, then use it instead of `Task.Delay`
 * Use it from tests